File: dart/2015/day03a/day03a.dart
#!/usr/bin/env dart

import 'dart:io';

import '../../utils/geometry/position2d.dart';

void usage(String progName) {
    print('usage: $progName <input file>');
    exit(1);
}

int process(String filename) {
    Position2D<int> santa = Position2D<int>(0, 0);
    var positions = <Position2D>{};
    positions.add(santa);

    final lines = File(filename).readAsLinesSync();
    for ( final line in lines ) {
        for ( final ch in line.runes ) {
            if ( ch == '^'.codeUnitAt(0) ) {
                santa = Position2D<int>(santa.x, santa.y + 1);
            } else if ( ch == 'v'.codeUnitAt(0) ) {
                santa = Position2D<int>(santa.x, santa.y - 1);
            } else if ( ch == '<'.codeUnitAt(0) ) {
                santa = Position2D<int>(santa.x - 1, santa.y);
            } else if ( ch == '>'.codeUnitAt(0) ) {
                santa = Position2D<int>(santa.x + 1, santa.y);
            }
            positions.add(santa);
        }
    }
    return positions.length;
}

void main(List<String> args) {
    if ( args.isEmpty ) {
        usage(Platform.script.pathSegments.last);
    }
    final filename = args[0];
    final result = process(filename);
    print('result = $result');
}

File: r/2015/day03a/day03a.r
#!/usr/bin/env Rscript

usage <- function() {
    args <- commandArgs()
    progname <- sub("^--file=", "", args[grep("^--file=", args)])
    cat("usage:", progname, "<input file>\n")
    quit(status = 1)
}

process <- function(filename) {
    positions <- list()
    santa <- c(0, 0)
    key <- paste(santa[1], santa[2], sep = ",")
    positions[[key]] <- TRUE
    lines <- suppressWarnings(readLines(filename))
    for ( line in lines ) {
        chars <- strsplit(line, "")[[1]]
        for ( ch in chars ) {
            if ( ch == '^' ) {
                santa[2] <- santa[2] + 1
            } else if ( ch == 'v' ) {
                santa[2] <- santa[2] - 1
            } else if ( ch == '<' ) {
                santa[1] <- santa[1] - 1
            } else if ( ch == '>' ) {
                santa[1] <- santa[1] + 1
            }
            key <- paste(santa[1], santa[2], sep = ",")
            positions[[key]] <- TRUE
        }
    }
    return(length(positions))
}

main <- function() {
    args <- commandArgs(trailingOnly = TRUE)
    if ( length(args) < 1 ) {
        usage()
    }
    filename <- args[1]
    result <- process(filename)
    cat("result =", result, "\n")
}

main()
